www/js/diary/services.js

```diff
@@ -406,6 +413,12 @@ angular.module('emission.main.diary.services', ['emission.plugin.logger',
   }
 
   dh.getUserInputForTrip = function(tripgj, nextTripgj, userInputList) {
+    // If there is only one item in the list, return it.
+    // This make it compatible when fake list is given (for Enketo Survey).
+    // As well as optimize the performance.
+    if (userInputList.length === 1) {
```

Comment:
Need this inorder for https://github.com/e-mission/e-mission-phone/pull/798/commits/fb338f12df472f002f4b591bc315d9cd0023fef7#diff-b254f964755cdc55e579741bb46be006886e57fa24e7302e413556ab39ccb014R492 to work.

Reply:
I see. But why do you need to call

```
$scope.populateInputFromTimeline(tripgj, undefined, inputType, [{data: result}])
```

from `openPopover` in the first place? 

`populateInputFromTimeline`, and by extension, `getUserInputForTrip` are used for matching the input to the trip. You don't really need that because you know the match already; it is the new trip.
I don't call `populateInputFromTimeline` for the trip confirm inputs

Reply:
Inside `openPopover`, we call `EnketoSurveyLaunch.launch` to launch the survey. Then when we receive the answer, it comes in the `.then()` clause of the promise. Only that time that we can get the answer and update the label for that specific trip.

Reply:
I understand how promises work. What I don't understand is why you need to call `populateInputFromTimeline`. For example, `populateInputFromTimeline` calls, in its first line:

```
        var userInput = DiaryHelper.getUserInputForTrip(tripgj, nextTripgj, inputList);
```

This iterates through all the user inputs and tries to match the trip. But in your case, you already know which trip this is the input for. So why take that additional performance hit instead of just setting the value into the trip directly?

e.g. something like

```
tripgj.userInput[inputType] = result;
```

or 

```
tripgj.userInput[inputType] = {data: result};
```

or something similar

As I said earlier, that's what I do with the current user labels (see `$scope.save` for example).

Reply:
hmm, haven't think of that before. Thank you for pointing out. Let me try...

Reply:
I only thought that putting all ui update in the same place would be nice.

Reply:
Done https://github.com/e-mission/e-mission-phone/pull/798/commits/c5ead07d23f4e18f33bcd2bfdea7943f9ddcf534